internal/services/connection/models.go

```diff
@@ -0,0 +1,141 @@
+	to.ID = customtypes.NewUUIDPointerValue(from.ID)
+	to.DisplayName = types.StringPointerValue(from.DisplayName)
+	to.GatewayID = customtypes.NewUUIDPointerValue(from.GatewayID)
+	to.ConnectivityType = types.StringPointerValue((*string)(from.ConnectivityType))
+	to.PrivacyLevel = types.StringPointerValue((*string)(from.PrivacyLevel))
+
+	var diags diag.Diagnostics
+
+	var connectionDetails supertypes.SingleNestedObjectValueOf[ConnectionDetails]
+	var connectionDetailsModel *ConnectionDetails
+
+	if to.ConnectionDetails.IsKnown() {
+		connectionDetails = to.ConnectionDetails
+
+		connectionDetailsModel, diags = to.ConnectionDetails.Get(ctx)
+		if diags.HasError() {
+			return diags
+		}
+	} else {
+		var defaultModel ConnectionDetails
+		connectionDetailsModel = &defaultModel
+	}
+
+	if from.ConnectionDetails != nil {
+		var connectionDetailsModelPtr *ConnectionDetails
+
+		switch v := any(connectionDetailsModel).(type) {
+		case *dsConnectionDetailsModel:
+			setDSConnectionDetails(*from.ConnectionDetails, v)
+
+			if convertedValue, ok := any(*v).(ConnectionDetails); ok {
+				connectionDetailsModelPtr = &convertedValue
+			}
+		case *rsConnectionDetailsModel:
+			setRSConnectionDetails(*from.ConnectionDetails, v)
+
+			if convertedValue, ok := any(*v).(ConnectionDetails); ok {
+				connectionDetailsModelPtr = &convertedValue
+			}
+		}
+
+		if diags := connectionDetails.Set(ctx, connectionDetailsModelPtr); diags.HasError() {
+			return diags
+		}
+	}
+
+	to.ConnectionDetails = connectionDetails
+
+	var credentialDetails supertypes.SingleNestedObjectValueOf[CredentialDetails]
+	var credentialDetailsModel *CredentialDetails
+
+	if to.CredentialDetails.IsKnown() {
+		credentialDetails = to.CredentialDetails
+
+		credentialDetailsModel, diags = to.CredentialDetails.Get(ctx)
+		if diags.HasError() {
+			return diags
+		}
+	} else {
+		credentialDetails = supertypes.NewSingleNestedObjectValueOfNull[CredentialDetails](ctx)
+		var defaultModel CredentialDetails
+		credentialDetailsModel = &defaultModel
+	}
+
+	if from.CredentialDetails != nil {
+		var credentialDetailsModelPtr *CredentialDetails
+
+		switch v := any(credentialDetailsModel).(type) {
+		case *dsCredentialDetailsModel:
+			setDSCredentialDetails(*from.CredentialDetails, v)
+
+			if convertedValue, ok := any(*v).(CredentialDetails); ok {
+				credentialDetailsModelPtr = &convertedValue
+			}
+		case *rsCredentialDetailsModel:
+			setRSCredentialDetails(*from.CredentialDetails, v)
+
+			if convertedValue, ok := any(*v).(CredentialDetails); ok {
+				credentialDetailsModelPtr = &convertedValue
+			}
+		}
+
+		if diags := credentialDetails.Set(ctx, credentialDetailsModelPtr); diags.HasError() {
+			return diags
+		}
+	}
+
+	to.CredentialDetails = credentialDetails
+
+	return nil
+}
+
+type credentialsBasicModel struct {
+	Username          types.String `tfsdk:"username"`
+	PasswordWO        types.String `tfsdk:"password_wo"`
```

Comment:
Where does this "wo" suffix that I see in several fields come from? It feels... weird and unnatural.